### PR TITLE
Close / Reopen issue context menu and keyboard shortcuts

### LINKED ISSUE #1295: Implement context menu + keyboard shortcuts for close / reopen issues

### PR DESCRIPTION
**Pull request by xinan**
Closes #1295 

Keyboard Shortcuts:
- Close Issue (C)
- Reopen Issue (O)

Screenshot:
screenshot

### CODE REVIEW

**Comment by damithc** [comment]
I don't think close and open should be done using one letter shortcuts as they can be triggered by mistake. At least add a confirmation dialog.

**Comment by damithc** [comment]
You may want to use X for close, because 'C' is used by GitHub as the shortcut for 'Compose'.

**Comment by xinan** [comment]
It seems that "X" is a prefix for another command, so I cannot use just it for closing issues. How about using a single "T" (for "Toggle") to do both closing and reopening?

**Comment by damithc** [comment]
`S` (for **s**tatus)? Is it being used for something else?

**Comment by xinan** [comment]
I don't quite get what you mean prof.

**Comment by damithc** [comment]
Use `s` as the keyboard shortcut to toggle status, but have a confirmation dialog to prevent accidental closure/reopening.

**Comment by xinan** [comment]
Oh I see. I figured out that we can use "X" though. Which one do you prefer? "X/O" or "S"?

**Comment by damithc** [comment]
Sure,  `X` is fine too. We can save `s` for something else.

**Comment by xinan** [comment]
Tests for it are added to `unstable`. Is that why the coverage dropped?

File: src/main/java/backend/Logic.java [hunk 1]
```diff
@@ -331,6 +331,30 @@ private void revertLocalLabelsReplace(TurboIssue modifiedIssue, List<String> ori
     }
 
     /**
+     * Edit the state of an issue on GitHub
+     *
+     * @param issue The issue whose state is to be updated
+     * @param open The new state for the issue
+     * @return True for success, false otherwise
+     */
+    public CompletableFuture<Boolean> editIssueState(TurboIssue issue, boolean open) {
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: Boolean variable should be `isOpen`

File: src/main/java/backend/Logic.java [hunk 2]
```diff
@@ -331,6 +331,30 @@ private void revertLocalLabelsReplace(TurboIssue modifiedIssue, List<String> ori
...
+        String action = open ? "Reopening" : "Closing";
+        logger.info(String.format("%s %s on GitHub", action, issue));
+        return repoIO.editIssueState(issue, open)
+                .thenApply(success -> {
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: Boolean variable should be `isSuccessful`

File: src/main/java/backend/Logic.java [hunk 3]
```diff
@@ -331,6 +331,30 @@ private void revertLocalLabelsReplace(TurboIssue modifiedIssue, List<String> ori
...
+                    if (success) {
+                        logger.info("Changing state for " + issue + " on UI");
+                        issue.setOpen(open);
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: It's unsafe to modify a dangly reference to `TurboIssue` to change the state of repositories. You can refer to `replaceIssueLables` method to see how we currently modify a local state. You can also take a look at the header comment of `MultiModel`

File: src/main/java/backend/Logic.java [hunk 4]
```diff
@@ -331,6 +331,30 @@ private void revertLocalLabelsReplace(TurboIssue modifiedIssue, List<String> ori
...
+                        refreshUI();
+                        return true;
+                    } else {
+                        return false;
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: The `else` clause here is not necessary. The preferred controlled flow is to deal with the exceptional cases first e.g.
```java
if (!isSuccessful) {
  return false;
}
// Do the main successful flow
....
```

File: src/main/java/backend/github/EditIssueStateTask.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package backend.github;
+
+import backend.interfaces.Repo;
+import backend.interfaces.TaskRunner;
+
+import java.io.IOException;
+
+public class EditIssueStateTask extends GitHubRepoTask<Boolean> {
```
Comment (ndt93) [inline src/main/java/backend/github/EditIssueStateTask.java]: Needs class header comment here.

File: src/main/java/ui/listpanel/ConfirmCloseReopenDialog.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package ui.listpanel;
+
+import javafx.scene.control.Alert;
+import javafx.stage.Stage;
+
+public class ConfirmCloseReopenDialog extends Alert {
```
Comment (ndt93) [inline src/main/java/ui/listpanel/ConfirmCloseReopenDialog.java]: Needs header comment.

File: src/main/java/ui/listpanel/ConfirmCloseReopenDialog.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    private static final String DIALOG_TITLE = "%s Issue?";
+    private static final String DIALOG_HEADER = "This will be %s on Github.";
+    private static final String DIALOG_CONTENT = "Are you sure?";
+
+    public ConfirmCloseReopenDialog(Stage stage, boolean open) {
+        super(AlertType.CONFIRMATION);
```
Comment (ndt93) [inline src/main/java/ui/listpanel/ConfirmCloseReopenDialog.java]: Boolean variable should be `isOpen`

File: src/main/java/ui/listpanel/ListPanel.java [hunk 8]
```diff
@@ -404,6 +421,35 @@ private MenuItem updateMarkAsReadUnreadMenuItem() {
...
+            closeReopenIssueMenuItem.setDisable(false);
+            TurboIssue selectedIssue = item.get().getIssue();
+
+            if (selectedIssue.isOpen()) {
+                closeReopenIssueMenuItem.setText(closeIssueMenuItemText);
+            } else {
+                closeReopenIssueMenuItem.setText(reopenIssueMenuItemText);
+            }
+        } else {
+            closeReopenIssueMenuItem.setDisable(true);
```
Comment (ndt93) [inline src/main/java/ui/listpanel/ListPanel.java]: This `else` clause should also come first, and you can also use `return` to reduce nesting `if/else`

**Comment by ndt93** [comment]
@xinan I think we need a number of unit tests to cover the backend changes at least, since the context menu tests are unstable. You can refer to existing tests of similar functionalities e.g. in `LogicTests`, `MultiModelTests`, `ModelTests`, `RepoOpControlTests` etc.

**Comment by ndt93** [comment]
@xinan Is this ready for review again?

**Comment by xinan** [comment]
@ndt93 Yes!

File: src/main/java/backend/Logic.java [hunk 9]
```diff
@@ -331,6 +331,74 @@ private void revertLocalLabelsReplace(TurboIssue modifiedIssue, List<String> ori
...
+        LocalDateTime originalStateModifiedAt = modifiedIssue.getStateLastModifiedAt();
+        LocalDateTime currentStateModifiedAt = currentIssue.getStateLastModifiedAt();
+        boolean isCurrentStateModifiedOriginalState = originalStateModifiedAt.isEqual(currentStateModifiedAt);
+
+        if (isCurrentStateModifiedOriginalState) {
+            logger.info("Reverting state for issue " + currentIssue);
+            models.editIssueState(currentIssue.getRepoId(), currentIssue.getId(), isOpenOriginally);
+            refreshUI();
+        }
+    }
```
Comment (damithc) [inline src/main/java/backend/Logic.java]: This is a general comment. Use nested blocks to handle abnormal branches. The normal flow of the method should be in the un-nested code.

File: src/main/java/backend/Logic.java [hunk 10]
```diff
@@ -394,6 +394,78 @@ private void revertLocalMilestoneReplace(TurboIssue originalIssue) {
     }
 
     /**
+     * Edit the state of an issue on GitHub
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: This comment doesn't look like it belong here...

File: src/main/java/backend/Logic.java [hunk 11]
```diff
@@ -394,6 +394,78 @@ private void revertLocalMilestoneReplace(TurboIssue originalIssue) {
...
+        }
+
+        logger.error("Unable to update model on server");
+        revertLocalStateEdit(locallyModifiedIssue.get(), isOpenOriginally);
+        return false;
+    }
+
+    /**
+     * Set the state of the issue in the {@link Logic#models} corresponding to {@code modifiedIssue} to
+     * {@code isOpenOriginally} if the current state on the issue is assigned at the same time as {@code modifiedIssue}
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: Should be `Sets the state...`

File: src/main/java/backend/Logic.java [hunk 12]
```diff
@@ -394,6 +394,78 @@ private void revertLocalMilestoneReplace(TurboIssue originalIssue) {
...
+     * @param isOpenOriginally
+     */
+    private void revertLocalStateEdit(TurboIssue modifiedIssue, boolean isOpenOriginally) {
+        TurboIssue currentIssue = getIssue(modifiedIssue.getRepoId(), modifiedIssue.getId()).orElse(modifiedIssue);
+        LocalDateTime originalStateModifiedAt = modifiedIssue.getStateLastModifiedAt();
+        LocalDateTime currentStateModifiedAt = currentIssue.getStateLastModifiedAt();
+        boolean isCurrentStateModifiedOriginalState = originalStateModifiedAt.isEqual(currentStateModifiedAt);
+
+        if (isCurrentStateModifiedOriginalState) {
+            return;
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: Should have some `log`s here to report this scenario

File: src/main/java/backend/RepoIO.java [hunk 13]
```diff
@@ -207,6 +207,10 @@ private Model downloadRepoFromSourceBlocking(String repoId) {
         return repoSource.replaceIssueMilestone(issue, milestone);
     }
 
+    public CompletableFuture<Boolean> editIssueState(TurboIssue issue, boolean open) {
```
Comment (ndt93) [inline src/main/java/backend/RepoIO.java]: Boolean variables should be `isOpen`. This should apply to call instances of `open` boolean parameter or variable in other parts.

File: src/main/java/undo/actions/EditIssueStateAction.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        this.open = open;
+    }
+
+    @Override
+    public String getDescription() {
+        if (open) {
+            return DESCRIPTION_REOPENED;
+        } else {
+            return DESCRIPTION_CLOSED;
+        }
```
Comment (ndt93) [inline src/main/java/undo/actions/EditIssueStateAction.java]: Can remove `else` clause here.

**Comment by ndt93** [comment]
Some cosmetic changes requested.

File: src/main/java/backend/Logic.java [hunk 15]
```diff
@@ -394,6 +394,78 @@ private void revertLocalMilestoneReplace(TurboIssue originalIssue) {
...
+        }
+
+        logger.info("Reverting state for issue " + currentIssue);
+        models.editIssueState(currentIssue.getRepoId(), currentIssue.getId(), isOpenOriginally);
+        refreshUI();
+    }
+
+    /**
+     * Edit the open/closed state in the issue object, the UI, and the server, in that order.
+     * Server update is done after the local update to reduce the lag between the user action and the UI response.
```
Comment (ndt93) [inline src/main/java/backend/Logic.java]: Should be `Edits`

File: src/main/java/backend/github/GitHubRepo.java [hunk 16]
```diff
@@ -281,6 +281,20 @@ public boolean login(UserCredentials credentials) {
                 .map(Milestone::getNumber);
     }
 
+    public boolean editIssueState(String repoId, int issueId, boolean isOpen) throws IOException {
+        Issue updatedIssue = issueService.editIssueState(
+                RepositoryId.createFromId(repoId),
+                issueId,
+                isOpen
+        );
+        // Return `true` upon success, or `false` otherwise.
```
Comment (ndt93) [inline src/main/java/backend/github/GitHubRepo.java]: This should be in the method's header comment instead.

File: src/main/java/backend/resource/Model.java [hunk 17]
```diff
@@ -231,6 +231,21 @@ private void ______OPERATIONS_____() {}
                 () -> logger.error("Issue " + issueId + " not found in model for " + repoId));
     }
 
+    /**
+     * Set the open/closed state of an issue specified by {@code issueId} with {@code isOpen}
+     *
```
Comment (ndt93) [inline src/main/java/backend/resource/Model.java]: Should be `Sets`

File: src/main/java/backend/resource/MultiModel.java [hunk 18]
```diff
@@ -121,6 +121,21 @@ public synchronized MultiModel replace(Model newModel) {
                 () -> logger.error("Model " + repoId + " not found in models"));
     }
 
+    /**
+     * Set the open/closed state of an issue specified by {@code issueId} in {@code repoId} to {@code isOpen}
+     *
```
Comment (ndt93) [inline src/main/java/backend/resource/MultiModel.java]: Should be `Sets`

File: src/main/java/backend/github/GitHubRepo.java [hunk 19]
```diff
@@ -281,6 +281,29 @@ public boolean login(UserCredentials credentials) {
...
+                RepositoryId.createFromId(repoId),
+                issueId,
+                isOpen
+        );
+
+        if (isOpen) {
+            return updatedIssue.getState().equals(IssueService.STATE_OPEN);
+        } else {
+            return updatedIssue.getState().equals(IssueService.STATE_CLOSED);
+        }
```
Comment (damithc) [inline src/main/java/backend/github/GitHubRepo.java]: How about this?
```java
return updatedIssue.getState().equals(isOpen?IssueService.STATE_OPEN:IssueService.STATE_CLOSED)
```

File: docs/keyboardShortcuts.md [hunk 20]
```diff
@@ -60,6 +60,8 @@ These shortcuts can be used when the focus is on an issue card in a panel.
 | Set **M**ilestone|<kbd>M</kbd>|Jump to the *milestone* property of the issue in the browser view|
 | Mark as R**e**ad|<kbd>E</kbd>|Mark an issue as read. Read issues will be shown in a different color and can be filtered. After marking an issue as *read*, it will remain in *read* state until the issue is updated.|
 | Mark as **U**nread|<kbd>U</kbd>|Reverts an issue's *read* status|
+| **C**lose Issue|<kbd>C</kbd>|Close an issue on GitHub|
+| Re**o**pen Issue|<kbd>O</kbd>|Reopen an closed issue on GitHub|
```
Comment (damithc) [inline docs/keyboardShortcuts.md]: Didn't we decide to use `X`?

Comment (xinan) [inline docs/keyboardShortcuts.md]: Oops sorry I forgot to update the documentation. It is `X` :smile:

File: src/main/java/backend/github/EditIssueStateTask.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+        this.issueId = issueId;
+        this.isOpen = isOpen;
+    }
+
+    @Override
+    public void run() {
+        try {
+            response.complete(
+                    repo.editIssueState(repoId, issueId, isOpen)
+            );
```
Comment (damithc) [inline src/main/java/backend/github/EditIssueStateTask.java]: Is the line wrapping necessary?

File: src/main/java/backend/Logic.java [hunk 22]
```diff
@@ -394,6 +394,78 @@ private void revertLocalMilestoneReplace(TurboIssue originalIssue) {
     }
 
     /**
+     * Handles the result of updating an issue's state on server. Current implementation includes
+     * reverting back to the original state locally if the server update failed.
```
Comment (damithc) [inline src/main/java/backend/Logic.java]: Get rid of unnecessary words. e.g. 
~~`Current implementation includes reverting back to the original state locally if the server update failed.`~~ 
--> `Reverts back to the original state locally if the server update failed.`

File: src/main/java/github/IssueServiceEx.java [hunk 23]
```diff
@@ -41,11 +41,11 @@ public Issue createIssue(IRepositoryIdProvider repository, Issue issue) throws I
         return returnedIssue;
     }
 
-    public Issue editIssueState(IRepositoryIdProvider repository, int issueId, boolean open) throws IOException {
+    public Issue editIssueState(IRepositoryIdProvider repository, int issueId, boolean isOpen) throws IOException {
         HttpURLConnection connection = createIssuePostConnection(repository, issueId);
         HashMap<Object, Object> data = new HashMap<>();
         String state;
-        if (open) {
+        if (isOpen) {
```
Comment (damithc) [inline src/main/java/github/IssueServiceEx.java]: While you are at it, just make it `state = isOpen ? ... : ... ` ?

File: src/main/java/ui/listpanel/ListPanel.java [hunk 24]
```diff
@@ -523,6 +574,29 @@ private void markAsUnread() {
         }
     }
 
+    private void confirmCloseReopen(GuiElement element, boolean isOpen) {
```
Comment (damithc) [inline src/main/java/ui/listpanel/ListPanel.java]: The method name is not very clear

**Comment by damithc** [comment]
Some cosmetic changes requested.

**Comment by damithc** [comment]
Can merge after addressing latest comments.

File: src/main/java/undo/actions/EditIssueStateAction.java [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+        this.isOpen = isOpen;
+    }
+
+    @Override
+    public String getDescription() {
+        if (isOpen) {
+            return DESCRIPTION_REOPENED;
+        }
+
+        return DESCRIPTION_CLOSED;
```
Comment (damithc) [inline src/main/java/undo/actions/EditIssueStateAction.java]: Collapse into one ternary statement?

**Comment by damithc** [comment]
Can merge after addressing the minor comment.

**Pull request merged**